Add functionality for the user to be able to copy/paste scenes and links

DanielSBrown opened the pull request:
Addresses #76 
This is a decent amount of WIP -- I have it working I think, but will definitely need to clean this up big time. 

Getting a PR out so that more experienced eyes can get on this. Full disclosure, a lot of this code smells really bad; tons of nested and sequential loops that could probably be slimmed down, which is what I'll be working on in the next commit.

averyvery commented on app/javascript/src/components/Editor.tsx:
```diff
@@ -19,6 +20,11 @@ import './FlowChart.css'
 import { save } from '../persistance'
 import { clone } from '../clone'
+import * as _ from 'lodash'
+import { link } from 'fs'
```
Maybe I'm misreading, but should this be importing from 'fs'? At first glance it looks like maybe your editor auto-imported this inappropriately; I don't see it used in the file.

averyvery commented on app/javascript/src/components/Editor.tsx:
```diff
@@ -363,6 +375,148 @@ class Editor extends React.Component<EditorProps, EditorState> {
+    for (let node of pastedNodes) {
+      model.addNode(node)
+    }
+    for (let link of pastedLinks) {
```
Could this be `pastedLinks.forEach(model.addLink)`? Maybe I just have a superstition about this, but `array.[functional method]()` tends to be nicer/clearer than for loops in JS.

averyvery commented on app/javascript/src/components/Editor.tsx:
```diff
@@ -363,6 +375,148 @@ class Editor extends React.Component<EditorProps, EditorState> {
+
+    const selectedNodes = model.getSelectedItems().filter(item => {
+      return item instanceof DefaultNodeModel
+    }) as DefaultNodeModel[]
```
Style thing, I think this could be cleaned up to use parens, which return their value:

`.filter(item => (
  item instanceOf DefaultNodeModel
)`

averyvery commented on app/javascript/src/components/Workspace.tsx:
```diff
@@ -54,9 +58,24 @@ export default class Workspace extends React.Component<WorkspaceProps> {
   }
 
+  handleKeyPress = (event: any) => {
+    if ((event.ctrlKey || event.metaKey) && event.keyCode == 67)
```
To save people the time it takes to look up keycodes, sometimes I like to write'em as consts:

```
const C_KEY = 67
const V_KEY = 86

...

event.keyCode = V_KEY
```

Probably just a style preference.

averyvery commented on app/javascript/src/components/Editor.tsx:
```diff
@@ -363,6 +375,148 @@ class Editor extends React.Component<EditorProps, EditorState> {
+    let ret = []
+    for (let node of copiedNodes) {
+      // Iterates over the nodes that were copied to clipboard and then over their ports, returning them for the pasted links to use
+      let copiedOutPorts = node.getOutPorts()
```
Just for my education (I've never worked on storybook), what are "ports" in this context, and where is `getOutPorts` coming from?

DanielSBrown replied:
<img width="471" alt="Screen Shot 2020-02-27 at 8 40 20 AM" src="https://user-images.githubusercontent.com/16651533/75450439-d08c8000-593c-11ea-8370-133b846460fd.png">

The ports are the connecting dots, with the inPorts being where any lines are coming into the node, and outPorts are the points where the lines leave a node. 

This codebase uses the [react-diagrams](https://github.com/projectstorm/react-diagrams) library, and `getOutPorts()` is a method on that library's `DefaultNodeModel` class, where it returns every port on that node that is an "out" port. In this app, I believe all out ports are always on the right side of the node by convention.

averyvery replied:
Ahhh, ok thanks! I googled "getOutPorts" and couldn't find any JS libs with it, but I see it now in react-diagrams 👍

averyvery replied:
Ahhh, ok thanks! I googled "getOutPorts" and couldn't find any JS libs with it, but I see it now in react-diagrams 👍

averyvery commented on app/javascript/src/components/Editor.tsx:
```diff
@@ -363,6 +375,148 @@ class Editor extends React.Component<EditorProps, EditorState> {
+
+    pastedLink.sourcePort = sourcePort
+    pastedLink.targetPort = targetPort
+    // These lines below seem redundant, but removing them causes the pasted link(s) not to move with the rest of the objects until page reload
```
:+1:

DanielSBrown commented:
@averyvery Addressed your feedback items, thanks a lot for them! I agree that the `array.[functional method]()` looks much cleaner, so I added that to additional places where it could be added in context of this story.

efatsi commented on app/javascript/src/components/Editor.tsx:
```diff
@@ -363,6 +372,143 @@ class Editor extends React.Component<EditorProps, EditorState> {
+          ret.push(pastedNodes[nodeIndex].getInPorts()[portIndex])
+        }
+      })
+    })
```
maybe not the most efficient code but looks like this certainly works!

efatsi commented on app/javascript/src/components/Editor.tsx:
```diff
@@ -363,6 +372,143 @@ class Editor extends React.Component<EditorProps, EditorState> {
+      .filter(item => item instanceof DefaultNodeModel) as DefaultNodeModel[]
+    const selectedLinks = this.model
+      .getSelectedItems()
+      .filter(item => item instanceof DefaultLinkModel) as DefaultLinkModel[]
```
scope creep feature request: If I shift-click on two linked scenes but don't select the joining link, it would be awesome if the joining link was added to the copiedLinks variable. I imagine another doubly nested loop with some complicated logic might do the trick?

DanielSBrown replied:
Added in the latest commit -- It wasn't quite as flat as just two loops, since the way to get Links from Nodes is to go through Ports, which ups the number of loops from having to loop through the ports first.
